src/main/java/org/apache/mesos/hdfs/Scheduler.java

```diff
@@ -76,15 +77,185 @@ public void registered(SchedulerDriver driver, FrameworkID frameworkId, MasterIn
+                log.info(String.format("NameNodes: %s", persistentState.getNameNodes()));
+                log.info(String.format("DataNodes: %s", persistentState.getDataNodes()));
+                correctCurrentPhase();
+                driver.declineOffer(offer.getId());
+              } else {
+                log.info("Declining offers while reconciling tasks");
+                driver.declineOffer(offer.getId());
+              }
+              break;
+            case JOURNAL_NODES :
+              if (liveState.getJournalNodeSize() < conf.getJournalNodeCount()) {
+                if (tryToLaunchJournalNode(driver, offer)) {
+                  acceptedOffer = true;
+                } else {
+                  driver.declineOffer(offer.getId());
+                }
+              }
+              break;
+            case START_NAME_NODES :
+              if (tryToLaunchNameNode(driver, offer)) {
+                acceptedOffer = true;
+              } else {
+                driver.declineOffer(offer.getId());
+              }
+              break;
+            case FORMAT_NAME_NODES :
+              if (tryToLaunchNameNode(driver, offer)) {
+                acceptedOffer = true;
+              } else {
+                driver.declineOffer(offer.getId());
+              }
+              break;
+            case DATA_NODES :
+              if (tryToLaunchDataNode(driver, offer)) {
+                acceptedOffer = true;
+              } else {
+                driver.declineOffer(offer.getId());
+              }
+              break;
+          }
+        }
+      }
+    }
+  }
+
+  @Override
+  public void slaveLost(SchedulerDriver driver, SlaveID slaveId) {
+    log.info("Slave lost slaveId=" + slaveId.getValue());
   }
 
+  @Override
+  public void run() {
+    FrameworkInfo.Builder frameworkInfo = FrameworkInfo.newBuilder()
+        .setName(conf.getFrameworkName())
+        .setFailoverTimeout(conf.getFailoverTimeout())
+        .setUser(conf.getHdfsUser())
+        .setRole(conf.getHdfsRole())
+        .setCheckpoint(true);
+
+    try {
+      FrameworkID frameworkID = persistentState.getFrameworkID();
+      if (frameworkID != null) {
+        frameworkInfo.setId(frameworkID);
+      }
+    } catch (InterruptedException | ExecutionException | InvalidProtocolBufferException e) {
+      throw new RuntimeException(e);
+    }
+
+    MesosSchedulerDriver driver = new MesosSchedulerDriver(this, frameworkInfo.build(),
+        conf.getMesosMasterUri());
+    driver.run().getValueDescriptor().getFullName();
+  }
   private void launchNode(SchedulerDriver driver, Offer offer,
-      String nodeName, List<String> taskNames, String executorName) {
+        String nodeName, List<String> taskNames, String executorName) {
     log.info(String.format("Launching node of type %s with tasks %s", nodeName,
         taskNames.toString()));
     String taskIdName = String.format("%s.%s.%d", nodeName, executorName,
```

Comment:
Using currentTimeMillis is fine for now, but consider a failover from one scheduler node to another where the times are out of sync, and the second scheduler happens to choose the same taskId as a task previously launched by the first scheduler. We'll just have to explain that we expect synchronized times between scheduler nodes. A more robust solution would be to have a taskId counter in the persistent state. TODO?